verifier/src/aux_verifier.rs

```diff
@@ -0,0 +1,60 @@
+use alloc::{string::ToString, vec::Vec};
+use crypto::{ElementHasher, RandomCoin, RandomCoinError};
+use math::FieldElement;
+use utils::Deserializable;
+
+/// A trait for generating the random elements required for constructing the auxiliary trace.
+pub trait AuxTraceVerifier {
+    type AuxRandElements<E: Send + Sync>;
+    type AuxProof: Deserializable;
+    type Error: ToString;
+
+    /// Generates the random elements required for constructing the auxiliary trace. Optionally,
+    /// verifies the auxiliary proof.
+    fn verify_aux_trace<E, Hasher>(
+        &self,
+        aux_proof: Option<Self::AuxProof>,
+        transcript: &mut impl RandomCoin<BaseField = E::BaseField, Hasher = Hasher>,
+    ) -> Result<Self::AuxRandElements<E>, Self::Error>
+    where
+        E: FieldElement,
+        Hasher: ElementHasher<BaseField = E::BaseField>;
+}
```

Comment:
I would rename this trait into `AuxVerifier` and the `verify_aux_trace()` method into just `verify()`.

But also, it feels like we are trying to do too much with this trait. If I'm understanding things correctly, we have 3 types of auxiliary random values:

1. Random values needed for constructing user-defined auxiliary columns. The number of these random values was previously specified by `num_aux_segment_rands` (and which I think we might need to bring back).
2. Random values needed for constructing Lagrange kernel columns. The number of these values depends on the trace length.
3. Random values needed for constructing/verifying auxiliary proof.

Currently, this method (and its prover equivalent `build_aux_trace()`) is responsible for generating all 3 types of random values. But I'm wondering if that's the right approach.

An alternative could be to let the library generate the first two types of random values (since the data for these is known via the AIR definition and trace length) and have this method worry only about the 3rd type. Concretely, this could look like so:

`AuxRandValues` would become a struct which could look like:
```Rust
pub struct AuxRandElements<E> {
    user_defined: Vec<E>, // probably a better name is needed here
    lagrange: Vec<E>,
}
```

In the `Air` trait, we'd have:
```Rust
fn get_aux_rand_elements<E, R>(
     &self,
     public_coin: &mut R,
 ) -> Result<AuxRandElements, RandomCoinError>
```
This would generate random values both for user-defined columns and Lagrange kernel columns as needed. This would also eliminate the need for the `Air::get_lagrange_rand_elements()` method.

On the prover side, we'd change `build_aux_trace()` method to look as follows:
```Rust
fn build_aux_trace<E>(
    &self,
    main_trace: &Self::Trace,
    aux_rand_elements: &AuxRandElements<E>,
    public_coin: &mut Self::RandomCoin,
) -> (ColMatrix<E>, ProverAuxProof<Self>)
```

Or, if `AuxProof` needs to be generated before the randomness for Lagrange kernel columns, we could split this into two methods:
```Rust
/// Builds AuxProof object drawing randomness from the public coin as needed.
fn generate_aux_proof<E>(
    &self,
    main_trace: &Self::Trace,
    public_coin: &mut Self::RandomCoin,
) -> ProverAuxProof<Self>

/// Builds auxiliary trace columns.
fn build_aux_trace<E>(
    &self,
    main_trace: &Self::Trace,
    aux_rand_elements: &AuxRandElements<E>,
) -> ColMatrix<E>
```

One the verifier side, `AuxVerifier::verify` would look something like this:
```Rust
fn verify_aux_trace<E, Hasher>(
   &self,
   aux_proof: Self::AuxProof,
   public_coin: &mut impl RandomCoin<BaseField = E::BaseField, Hasher = Hasher>,
) -> Result<(), Self::Error>
```
And would be responsible only for verifying the `AuxProof`.

It is possible I've missed some crucial detail somewhere and this doesn't work for some reason. But if not, it feels to me like a cleaner approach.

Reply:
I agree that `build/verify_aux_trace()` needs to be improved, but came to a different conclusion.

> But also, it feels like we are trying to do too much with this trait. If I'm understanding things correctly, we have 3 types of auxiliary random values:

This isn't quite right. At the lowest level, the `Air` can't generate Lagrange kernel random values separately from building the GKR proof, as they are generated at some point in that process. We need to think of all Lagrange-related things in the repo as GKR-related things. Or specifically, the GKR instantiation of `AuxProof`. So concretely speaking, your random values of type 2 (Lagrange-kernel random values) are a subset of type 3 ("aux proof"-random values), but only in the case where the `AuxProof` is GKR.

I think this hints at a larger issue: we started explicitly supporting GKR in winterfell by adding the Lagrange kernel things, but then stopped halfway and made the `AuxProof` machinery generic over any auxiliary proof. I believe this explains most of the awkwardness of the current solution you're trying to fix here. If we moved the GKR implementation in winterfell, then it would enable us to make things a lot cleaner:

- We would get rid of `AuxProof`
  - The only `AuxProof` supported would be GKR
- Your suggestion for the `AuxRandElements` struct that separates the random values between "user" and `lagrange` would work well. The library would only ever need to know about `lagrange` random values (as an option), and nothing else, since we don't support any other type of auxiliary proof.
  - So we'd get rid of the `AuxRandElements` associated types too
- Get rid of `AuxTraceVerifier`
  - The `Air` would know whether or not GKR is being used, and would be either apply the GKR `AuxTraceVerifier`, or the default one that only generates `num_aux_rand_elements` "user" random values.
- etc.

It seems to me like we'd be able to drastically reduce the complexity of the implementation, and still be in line with our initial decision to explicitly supporting accelerating LogUp with GKR (with our introducing the Lagrange-specific machinery). It isn't clear to me anyway that the library is currently general/flexible enough to allow for other types of `AuxProof` out of the box. The only example we currently have (GKR) wasn't, since it needed the Lagrange kernel-specific machinery to work.

### My thoughts on next steps
- I would leave the implementation mostly as-is, since it works, and we'll be able to test out the GKR acceleration in the VM more quickly,
- If we like the results, and decide to switch the VM to using GKR for LogUp, then we move the GKR implementation in winterfell, and simplify the implementation as explained above.

Reply:
> This isn't quite right. At the lowest level, the `Air` can't generate Lagrange kernel random values separately from building the GKR proof, as they are generated at some point in that process.

Ah! I see - so, I was incorrect here. My current understanding of the process is as follows:

1. After the main trace is built, we build the auxiliary proof.
2. In the process of building this proof, we generate random values needed for Lagrange kernel column.
3. We may also have some "user defined" aux rand values, which we can draw from the random coin after the auxiliary proof is generated (we could do it before as well, I think - as long as we remain consistent between prover and verifier)
4. We then use these random values to build the auxiliary trace (which would currently include building Largange kernel column "by hand" if needed).
5. We also use the randomness from step 2 and 3 during constraint evaluation.

> I think this hints at a larger issue: we started explicitly supporting GKR in winterfell by adding the Lagrange kernel things, but then stopped halfway and made the `AuxProof` machinery generic over any auxiliary proof. I believe this explains most of the awkwardness of the current solution you're trying to fix here.

I agree with this assessment. And thinking about it now, it seems to me that we can make some additional assumptions which would make the current implementation less general but at the same time, will be less complex and a smaller "diff" with the existing code.

Here is what I'm thinking:

First, I still think we should have two separate methods: `generate_aux_proof()` and `build_aux_trace()`. `generate_aux_proof()` could look like so:
```Rust
fn generate_aux_proof<E>(
    &self,
    main_trace: &Self::Trace,
    public_coin: &mut Self::RandomCoin,
) -> (ProverAuxProof<Self>, Vec<E>)
```
The returned `Vec<E>` would be explicitly interpreted as random values needed for Lagrange kernel column (we could also wrap them in a newtype - e.g., `LagrangeRandElements`).

After this function returns, we would build `AuxRandElements` struct which would be similar to the one in my previous comment.
```Rust
pub struct AuxRandElements<E> {
    user_defined: Vec<E>, // probably a better name is needed here
    lagrange: Vec<E>,
}
```
To build this struct, we could use `Air::build_aux_rand_elements()` which could look like so:
```Rust
fn get_aux_rand_elements<E, R>(
    &self,
    public_coin: &mut R,
    lagrange_rand_elements: Vec<E>,
) -> Result<Vec<E>, RandomCoinError>
```
This would work similarly to how `get_aux_trace_segment_random_elements()` works in the existing code and would also eliminate the need for a separate `Air::get_lagrange_rand_elements()` introduced in this PR.

Then, these `AuxRandElements` would be passed to the auxiliary trace constructor. The function could look as:
```Rust
fn build_aux_trace<E>(
    &self,
    main_trace: &Self::Trace,
    aux_rand_elements: &AuxRandElements<E>,
) -> ColMatrix<E>
```
We could actually move this function back into the `Trace` trait - but I'm also fine with keeping it in the `Prover` trait.

**Assumption**: one assumption I'm making here is that we are not duplicating a lot of work between `generate_aux_proof()` and `build_aux_trace()`. If we are, then it may make sense to rethink this - but also, this could be an optimization to make later.

One the verifier side, we would make analogous changes:

We'd have `AuxVerifier::verify()` function which could look like so:
```Rust
fn verify<E, Hasher>(
   &self,
   aux_proof: Self::AuxProof,
   public_coin: &mut impl RandomCoin<BaseField = E::BaseField, Hasher = Hasher>,
) -> Result<Vec<E>, Self::Error>
```
Where the returned `Vec<E>` would again be assumed to explicitly contain random values for the Lagrange column.

Then we'd use the returned values to construct `AuxRandElements` via `Air:: get_aux_rand_elements()` and the rest of the process would work as it does now.

I would also probably put the `AuxVerifier` as an associated type on the `Air` trait. So, basically, the two associated types that we'd need to add to the `Air` trait would be as follows:

```Rust
/// An auxiliary (i.e. non-STARK) proof object. If not needed, set to `()`.
type AuxProof: Serializable + Deserializable;

/// An auxiliary (i.e., non-STARK) verifier for verifying auxiliary proofs. If not needed, set to `()`.
type AuxVerifier: AuxVerifier<AuxProof = Self::AuxProof>;
```
I think this should allow us to go back to a single `verify()` function in the verifier.

If the above works, my intuition is that it will reduce the size of this PR somewhat, will position us better to transition to integrating GRK proofs into Winterfell, and will also introduce fewer API changes for this transient state of the library.

Reply:
> We may also have some "user defined" aux rand values, which we can draw from the random coin after the auxiliary proof is generated (we could do it before as well, I think - as long as we remain consistent between prover and verifier)

I think "after" makes more sense, as this will cover:

1. The case when there is a Lagrange kernel: the subsequent drawn randomness will be used to construct the second auxiliary column alongside the Lagrange kernel column.
2. The case when there is no Lagrange kernel: the `generate_aux_proof` will be trivial and the subsequent drawn randomness will default to the auxiliary trace randomness as it currently exists.